File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/TripRowSequencePlanner.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.model.Paddock
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.roundToInt

/**
 * Row/path traversal pattern, mirroring the iOS `TrackingPattern` enum. The
 * raw value matches the iOS string stored in `trips.tracking_pattern` so the
 * two platforms stay interchangeable.
 */
enum class TrackingPattern(val rawValue: String) {
    SEQUENTIAL("sequential"),
    EVERY_SECOND_ROW("everySecondRow"),
    FIVE_THREE("fiveThree"),
    UP_AND_BACK("upAndBack"),
    TWO_ROW_UP_BACK("twoRowUpBack"),
    CUSTOM("custom"),

    /**
     * Free Drive — operator is not following a planned row sequence. No
     * sequence is generated; rows are detected live from GPS position.
     */
    FREE_DRIVE("freeDrive");

    val title: String
        get() = when (this) {
            SEQUENTIAL -> "One After Another"
            EVERY_SECOND_ROW -> "Every Second Row"
            FIVE_THREE -> "3/5 Pattern"
            UP_AND_BACK -> "Up and Back"
            TWO_ROW_UP_BACK -> "2 Row Up & Back"
            CUSTOM -> "Custom"
            FREE_DRIVE -> "Free Drive"
        }

    val subtitle: String
        get() = when (this) {
            SEQUENTIAL -> "0.5, 1.5, 2.5 ... lastRow+0.5"
            EVERY_SECOND_ROW -> "0.5, 2.5, 4.5 ... then 3.5, 1.5"
            FIVE_THREE -> "0.5, 3.5, 1.5, 5.5, 2.5 ... +5, -3"
            UP_AND_BACK -> "0.5, 0.5, 1.5, 1.5 ... each path twice"
            TWO_ROW_UP_BACK -> "Spray 2, Skip 2 up then back"
            CUSTOM -> "Save & reuse your own sequences"
            FREE_DRIVE -> "No planned path — detect rows live"
        }

    /**
     * Generate the local (1-based) path sequence for this pattern. Mirrors the
     * iOS `TrackingPattern.generateSequence`. Free Drive returns an empty list.
     */
    fun generateSequence(startRow: Int, totalRows: Int, reversed: Boolean = false): List<Double> {
        if (this == FREE_DRIVE) return emptyList()
        if (totalRows <= 0) return emptyList()

        val firstPath = startRow.toDouble() - 0.5
        val totalPaths = totalRows + 1

        var result: List<Double> = when (this) {
            SEQUENTIAL, CUSTOM -> (0 until totalPaths).map { firstPath + it.toDouble() }

            EVERY_SECOND_ROW -> {
                val sequence = mutableListOf<Double>()
                var i = 0
                while (i < totalPaths) {
                    sequence.add(firstPath + i.toDouble())
                    i += 2
                }
                i = totalPaths - 1
                while (i >= 0) {
                    val path = firstPath + i.toDouble()
                    if (!sequence.contains(path)) sequence.add(path)
                    i -= 2
                }
                for (j in 0 until totalPaths) {
                    val path = firstPath + j.toDouble()
                    if (!sequence.contains(path)) sequence.add(path)
                }
                sequence
            }

            UP_AND_BACK -> {
                val sequence = mutableListOf<Double>()
                for (j in 0 until totalPaths) {
                    val path = firstPath + j.toDouble()
                    sequence.add(path)
                    sequence.add(path)
                }
                sequence
            }

            FIVE_THREE -> generateFiveThreeSequence(firstPath, totalPaths)

            TWO_ROW_UP_BACK -> generateTwoRowUpBackSequence(firstPath, totalPaths)

            FREE_DRIVE -> emptyList()
        }

        if (reversed) result = result.reversed()
        return result
    }

    companion object {
        /** Decode a raw column value, defaulting to [SEQUENTIAL] like iOS. */
        fun fromRaw(raw: String?): TrackingPattern =
            entries.firstOrNull { it.rawValue == raw } ?: SEQUENTIAL

        private fun generateTwoRowUpBackSequence(firstPath: Double, totalPaths: Int): List<Double> {
            val lastPath = firstPath + (totalPaths - 1).toDouble()

            val upPaths = mutableListOf<Double>()
            var path = firstPath + 1.0
            while (path <= lastPath) {
                upPaths.add(path)
                path += 4.0
            }

            val backPaths = mutableListOf<Double>()
            path = firstPath + 3.0
            while (path <= lastPath) {
                backPaths.add(path)
                path += 4.0
            }
            backPaths.reverse()

            val result = (upPaths + backPaths).toMutableList()

            val visited = result.toSet()
            var r = firstPath
            while (r <= lastPath) {
                if (!visited.contains(r)) result.add(r)
                r += 1.0
            }
            return result
        }

        private fun generateFiveThreeSequence(firstPath: Double, totalPaths: Int): List<Double> {
            val startRow = firstPath
            val endRow = firstPath + (totalPaths - 1).toDouble()
            val startupPattern = listOf(0.0, 3.0, 1.0, 5.0, 2.0).map { startRow + it }

            val result = mutableListOf<Double>()
            val visited = mutableSetOf<Double>()

            for (row in startupPattern) {
                if (row in startRow..endRow && !visited.contains(row)) {
                    result.add(row)
                    visited.add(row)
                }
            }

            var current = result.lastOrNull()
                ?: return (0 until totalPaths).map { firstPath + it.toDouble() }

            var usePlusFive = true
            while (true) {
                val next = if (usePlusFive) current + 5.0 else current - 3.0
                if (next < startRow || next > endRow || visited.contains(next)) break
                result.add(next)
                visited.add(next)
                current = next
                usePlusFive = !usePlusFive
            }

            val allRows = mutableListOf<Double>()
            var row = startRow
            while (row <= endRow) {
                allRows.add(row)
                row += 1.0
            }
            val remaining = allRows.filter { !visited.contains(it) }
            if (remaining.isNotEmpty()) result.addAll(remaining)

            return result
        }
    }
}

/**
 * Shared row/path sequence planning, ported from the iOS
 * `TripRowSequencePlanner`. Encapsulates the multi-block aware path math so the
 * Start Path picker, Sequence Direction handling, and Proposed Row Sequence all
 * match iOS for a given selection. Real vineyard row numbers are preserved
 * (e.g. a block with rows 69–108 yields paths 68.5–108.5).
 *
 * Pure helper — no I/O, no Android dependencies — so it is unit-testable.
 */
object TripRowSequencePlanner {

    private fun rowsOf(paddock: Paddock): List<Int> =
        paddock.rows.orEmpty().map { it.number }

    /**
     * Comparator equivalent of the iOS `rowOrderSort`: lowest row number first,
     * then name; blocks without rows sort to the end (by name).
     */
    val rowOrderComparator: Comparator<Paddock> = Comparator { a, b ->
        val aMin = rowsOf(a).minOrNull()
        val bMin = rowsOf(b).minOrNull()
        when {
            aMin != null && bMin != null ->
                if (aMin != bMin) aMin.compareTo(bMin)
                else a.name.lowercase().compareTo(b.name.lowercase())
            aMin != null && bMin == null -> -1
            aMin == null && bMin != null -> 1
            else -> a.name.lowercase().compareTo(b.name.lowercase())
        }
    }

    /** Sorted, de-duplicated list of every actual row number across the blocks. */
    fun selectedRowNumbers(paddocks: List<Paddock>): List<Int> {
        val set = mutableSetOf<Int>()
        for (paddock in paddocks) {
            for (n in rowsOf(paddock)) set.add(n)
        }
        return set.sorted()
    }

    /** Combined total row count across every selected block. */
    fun combinedTotalRows(paddocks: List<Paddock>): Int =
        paddocks.sumOf { rowsOf(it).size }

    /** Whether any selected block has row geometry. */
    fun hasAnyRowGeometry(paddocks: List<Paddock>): Boolean =
        paddocks.any { rowsOf(it).isNotEmpty() }

    /**
     * Available start paths across the full selection. For each actual row
     * number N we contribute paths N-0.5 and N+0.5.
     */
    fun availablePaths(paddocks: List<Paddock>): List<Double> {
        val numbers = selectedRowNumbers(paddocks)
        if (numbers.isEmpty()) return listOf(0.5)
        val set = mutableSetOf<Double>()
        for (n in numbers) {
            set.add(n.toDouble() - 0.5)
            set.add(n.toDouble() + 0.5)
        }
        return set.sorted()
    }

    /** Clamp + snap a start path to a valid value within [availablePaths]. */
    fun clampedStartPath(value: Double, paddocks: List<Paddock>): Double {
        val paths = availablePaths(paddocks)
        val first = paths.firstOrNull() ?: return value
        val last = paths.lastOrNull() ?: return value
        var p = value
        if (p < first) p = first
        if (p > last) p = last
        val rounded = (p - 0.5).roundToInt().toDouble() + 0.5
        if (abs(rounded - p) > 0.01) {
            p = minOf(maxOf(rounded, first), last)
        }
        return p
    }

    /** Default start path — the very first path of the selection. */
    fun defaultStartPath(paddocks: List<Paddock>): Double =
        availablePaths(paddocks).firstOrNull() ?: 0.5

    /** Format a path with a single decimal where required ("1", "1.5"). */
    fun formatPath(value: Double): String =
        if (value % 1.0 == 0.0) value.toInt().toString()
        else String.format("%.1f", value)

    /** Friendly label for a path in a picker menu. */
    fun pathMenuLabel(path: Double, paddocks: List<Paddock>): String {
        val pathStr = formatPath(path)
        val numbers = selectedRowNumbers(paddocks)
        val lo = numbers.firstOrNull() ?: 1
        val hi = numbers.lastOrNull() ?: 0
        if (numbers.isNotEmpty() && path < lo.toDouble()) {
            return "Path before row $lo — $pathStr"
        }
        if (numbers.isNotEmpty() && path > hi.toDouble()) {
            return "Path after row $hi — $pathStr"
        }
        val lower = floor(path).toInt()
        val upper = lower + 1
        return "Between rows $lower\u2013$upper — $pathStr"
    }

    /** Combined row range label like "Rows 69–108" or "Rows not configured". */
    fun combinedRangeLabel(paddocks: List<Paddock>): String {
        val numbers = selectedRowNumbers(paddocks)
        val lo = numbers.firstOrNull()
        val hi = numbers.lastOrNull()
        if (lo == null || hi == null) return "Rows not configured"
        if (lo == hi) return "Row $lo"
        return "Rows $lo\u2013$hi"
    }

    /** Combined paths label like "Paths 0.5–108.5". */
    fun combinedPathsLabel(paddocks: List<Paddock>): String {
        val paths = availablePaths(paddocks)
        val lo = paths.firstOrNull() ?: return ""
        val hi = paths.lastOrNull() ?: return ""
        return "Paths ${formatPath(lo)}\u2013${formatPath(hi)}"
    }

    /** Compact row-range label from a sorted list of actual row numbers. */
    fun compactRowRangeLabel(numbers: List<Int>): String {
        val lo = numbers.firstOrNull() ?: return "Rows"
        val hi = numbers.lastOrNull() ?: return "Rows"
        val segments = mutableListOf<Pair<Int, Int>>()
        var segStart = lo
        var prev = lo
        for (n in numbers.drop(1)) {
            if (n == prev + 1) {
                prev = n
            } else {
                segments.add(segStart to prev)
                segStart = n
                prev = n
            }
        }
        segments.add(segStart to prev)
        if (segments.size == 1) {
            return if (lo == hi) "Row $lo" else "Rows $lo\u2013$hi"
        }
        if (segments.size <= 2) {
            val parts = segments.map { if (it.first == it.second) "${it.first}" else "${it.first}\u2013${it.second}" }
            return "Rows " + parts.joinToString(", ")
        }
        return "Rows $lo\u2013$hi"
    }

    /**
     * Generate the full traversal sequence for the given pattern, start path,
     * and direction across all selected paddocks. Returns paths expressed as
     * real row numbers (e.g. 68.5, 69.5, …), matching `Trip.rowSequence`.
     */
    fun generateSequence(
        paddocks: List<Paddock>,
        pattern: TrackingPattern,
        startPath: Double,
        directionHigherFirst: Boolean,
    ): List<Double> {
        val n = combinedTotalRows(paddocks)
        if (n <= 0) return emptyList()
        if (pattern == TrackingPattern.EVERY_SECOND_ROW) {
            return everySecondRowSequence(
                paths = availablePaths(paddocks),
                startPath = startPath,
                higherFirst = directionHigherFirst,
            )
        }
        val numbers = selectedRowNumbers(paddocks)
        val minRow = numbers.firstOrNull() ?: 1
        val offset = (minRow - 1).toDouble()
        val localStartRow = maxOf(1, minOf(((startPath - offset) + 0.5).toInt(), n))
        val raw = pattern.generateSequence(
            startRow = localStartRow,
            totalRows = n,
            reversed = !directionHigherFirst,
        )
        return raw.map { it + offset }
    }

    /** Every Second Row, parity-preserving sequence across the combined paths. */
    fun everySecondRowSequence(
        paths: List<Double>,
        startPath: Double,
        higherFirst: Boolean,
    ): List<Double> {
        if (paths.isEmpty()) return emptyList()
        val sorted = paths.sorted()
        val sameParity = sorted.filter { p ->
            val diff = (p - startPath).roundToInt()
            diff % 2 == 0
        }
        if (sameParity.isEmpty()) return emptyList()
        return if (higherFirst) {
            val firstRun = sameParity.filter { it >= startPath }.sorted()
            val wrap = sameParity.filter { it < startPath }.sorted()
            firstRun + wrap
        } else {
            val firstRun = sameParity.filter { it <= startPath }.sortedDescending()
            val wrap = sameParity.filter { it > startPath }.sortedDescending()
            firstRun + wrap
        }
    }

    /** Short preview text for the Proposed Row Sequence card. */
    fun sequencePreviewText(sequence: List<Double>, maxItems: Int = 10): String {
        val preview = sequence.take(maxItems).map { formatPath(it) }
        val joined = preview.joinToString(" \u2192 ")
        return if (sequence.size > maxItems) "$joined \u2192 \u2026" else joined
    }

    /** Human-readable explanation of how a pattern walks the paths; null for Free Drive. */
    fun patternPreviewNote(pattern: TrackingPattern): String? = when (pattern) {
        TrackingPattern.SEQUENTIAL ->
            "Sequential: walks every path one-by-one in the chosen direction."
        TrackingPattern.EVERY_SECOND_ROW ->
            "Every Second Row: advances by +2 in the chosen direction, then wraps to cover the remaining same-parity paths."
        TrackingPattern.FIVE_THREE ->
            "5/3 pattern: skips ahead 5, back 3, repeating from the chosen start."
        TrackingPattern.UP_AND_BACK ->
            "Up and Back: traverses then reverses, covering each path once."
        TrackingPattern.TWO_ROW_UP_BACK ->
            "Two Row Up & Back: pairs of rows, advancing then returning."
        TrackingPattern.CUSTOM ->
            "Custom pattern: generated from the chosen start and direction."
        TrackingPattern.FREE_DRIVE -> null
    }
}

File: ios/VineTrack/LegacyImported/Services/TripRowSequencePlanner.swift
import Foundation

/// Shared row/path sequence planning used by both Maintenance Trips
/// (`StartTripSheet`) and Spray Trips (`SprayCalculatorView`).
///
/// Encapsulates the multi-block aware path math so both flows produce the
/// **same** Start Path picker, Sequence Direction handling, and Proposed
/// Row Sequence for a given selection. Real vineyard row numbers are
/// preserved (e.g. a block with rows 69–108 yields paths 68.5–108.5).
enum TripRowSequencePlanner {
    /// Sort blocks by lowest row number first, then by name. Blocks without
    /// row geometry sort to the end (compared by name).
    static func rowOrderSort(_ a: Paddock, _ b: Paddock) -> Bool {
        let aMin = a.rows.map(\.number).min()
        let bMin = b.rows.map(\.number).min()
        switch (aMin, bMin) {
        case let (l?, r?):
            if l != r { return l < r }
            return a.name.lowercased() < b.name.lowercased()
        case (_?, nil): return true
        case (nil, _?): return false
        case (nil, nil): return a.name.lowercased() < b.name.lowercased()
        }
    }

    /// Sorted, de-duplicated list of every actual row number contributed by
    /// the selected paddocks. Drives the multi-block path range.
    static func selectedRowNumbers(in paddocks: [Paddock]) -> [Int] {
        var set = Set<Int>()
        for paddock in paddocks {
            for row in paddock.rows { set.insert(row.number) }
        }
        return set.sorted()
    }

    /// Combined total row count across every selected block.
    static func combinedTotalRows(in paddocks: [Paddock]) -> Int {
        paddocks.reduce(0) { $0 + $1.rows.count }
    }

    /// Whether any selected block has row geometry.
    static func hasAnyRowGeometry(_ paddocks: [Paddock]) -> Bool {
        paddocks.contains { !$0.rows.isEmpty }
    }

    /// Available start paths across the full selection. For each actual row
    /// number N we contribute paths N-0.5 (before/between) and N+0.5 (after).
    static func availablePaths(in paddocks: [Paddock]) -> [Double] {
        let numbers = selectedRowNumbers(in: paddocks)
        guard !numbers.isEmpty else { return [0.5] }
        var set = Set<Double>()
        for n in numbers {
            set.insert(Double(n) - 0.5)
            set.insert(Double(n) + 0.5)
        }
        return set.sorted()
    }

    /// Clamp + snap a start path to a valid value within `availablePaths`.
    static func clampedStartPath(_ value: Double, paddocks: [Paddock]) -> Double {
        let paths = availablePaths(in: paddocks)
        guard let first = paths.first, let last = paths.last else { return value }
        var p = value
        if p < first { p = first }
        if p > last { p = last }
        let rounded = (p - 0.5).rounded() + 0.5
        if abs(rounded - p) > 0.01 {
            p = min(max(rounded, first), last)
        }
        return p
    }

    /// Default start path when nothing has been chosen — the very first path
    /// of the selection (e.g. `minRow - 0.5`).
    static func defaultStartPath(in paddocks: [Paddock]) -> Double {
        availablePaths(in: paddocks).first ?? 0.5
    }

    /// Format a path value with a single decimal where required ("1", "1.5").
    static func formatPath(_ value: Double) -> String {
        if value.truncatingRemainder(dividingBy: 1) == 0 {
            return String(format: "%.0f", value)
        }
        return String(format: "%.1f", value)
    }

    /// Friendly label for a path in a picker menu, e.g.
    /// "Path before row 69 — 68.5" or "Between rows 12–13 — 12.5".
    static func pathMenuLabel(_ path: Double, paddocks: [Paddock]) -> String {
        let pathStr = formatPath(path)
        let numbers = selectedRowNumbers(in: paddocks)
        let lo = numbers.first ?? 1
        let hi = numbers.last ?? 0
        if !numbers.isEmpty, path < Double(lo) {
            return "Path before row \(lo) — \(pathStr)"
        }
        if !numbers.isEmpty, path > Double(hi) {
            return "Path after row \(hi) — \(pathStr)"
        }
        let lower = Int(floor(path))
        let upper = lower + 1
        return "Between rows \(lower)\u{2013}\(upper) — \(pathStr)"
    }

    /// Combined row range label like "Rows 69–108" or "Rows not configured".
    static func combinedRangeLabel(_ paddocks: [Paddock]) -> String {
        let numbers = selectedRowNumbers(in: paddocks)
        guard let lo = numbers.first, let hi = numbers.last else {
            return "Rows not configured"
        }
        if lo == hi { return "Row \(lo)" }
        return "Rows \(lo)\u{2013}\(hi)"
    }

    /// Combined paths label like "Paths 0.5–108.5".
    static func combinedPathsLabel(_ paddocks: [Paddock]) -> String {
        let paths = availablePaths(in: paddocks)
        guard let lo = paths.first, let hi = paths.last else { return "" }
        return "Paths \(formatPath(lo))\u{2013}\(formatPath(hi))"
    }

    /// Build a compact row-range label from a sorted list of actual row numbers.
    static func compactRowRangeLabel(_ numbers: [Int]) -> String {
        guard let lo = numbers.first, let hi = numbers.last else { return "Rows" }
        var segments: [(Int, Int)] = []
        var segStart = lo
        var prev = lo
        for n in numbers.dropFirst() {
            if n == prev + 1 {
                prev = n
            } else {
                segments.append((segStart, prev))
                segStart = n
                prev = n
            }
        }
        segments.append((segStart, prev))
        if segments.count == 1 {
            return lo == hi ? "Row \(lo)" : "Rows \(lo)\u{2013}\(hi)"
        }
        if segments.count <= 2 {
            let parts = segments.map { $0.0 == $0.1 ? "\($0.0)" : "\($0.0)\u{2013}\($0.1)" }
            return "Rows " + parts.joined(separator: ", ")
        }
        return "Rows \(lo)\u{2013}\(hi)"
    }

    /// Generate the full traversal sequence for the given pattern, start path,
    /// and direction across all selected paddocks. Returns paths expressed as
    /// real row numbers (e.g. `68.5`, `69.5`, …), matching what `Trip.rowSequence`
    /// expects.
    ///
    /// Every produced value is a member of `availablePaths(in:)` — the pattern
    /// walks *indices into that set*, so it can never invent a path above or
    /// below the selection, and never fabricates paths through row-number gaps
    /// in a non-contiguous multi-block selection.
    static func generateSequence(
        paddocks: [Paddock],
        pattern: TrackingPattern,
        startPath: Double,
        directionHigherFirst: Bool
    ) -> [Double] {
        guard combinedTotalRows(in: paddocks) > 0 else { return [] }
        return plannedSequence(
            paths: availablePaths(in: paddocks),
            pattern: pattern,
            startPath: startPath,
            higherFirst: directionHigherFirst
        )
    }

    /// The authoritative planner. `paths` is the universe of valid paths; the
    /// returned sequence is drawn entirely from it.
    ///
    /// Patterns operate on *indices* into the direction-ordered, start-rotated
    /// universe rather than on arithmetic over path values, so `startPath + 5`
    /// can never name a path that does not exist.
    static func plannedSequence(
        paths: [Double],
        pattern: TrackingPattern,
        startPath: Double,
        higherFirst: Bool
    ) -> [Double] {
        guard pattern != .freeDrive, !paths.isEmpty else { return [] }

        // Every Second Row already planned against the explicit valid path set;
        // that behaviour is preserved verbatim.
        if pattern == .everySecondRow {
            return everySecondRowSequence(
                paths: paths,
                startPath: startPath,
                higherFirst: higherFirst
            )
        }

        let ordered = orderedTraversal(paths: paths, startPath: startPath, higherFirst: higherFirst)
        guard !ordered.isEmpty else { return [] }

        switch pattern {
        case .sequential, .custom:
            return ordered
        case .upAndBack:
            return ordered.flatMap { [$0, $0] }
        case .fiveThree:
            return indexed(ordered, by: fiveThreeIndices(count: ordered.count))
        case .twoRowUpBack:
            return indexed(ordered, by: twoRowUpBackIndices(count: ordered.count))
        case .everySecondRow, .freeDrive:
            return ordered
        }
    }

    /// The valid path universe ordered by traversal direction and rotated so
    /// the selected start path is first.
    ///
    /// `higherFirst` mirrors the picker: `true` = "Lower to higher" (ascending),
    /// `false` = "Higher to lower" (descending). Rotation — rather than
    /// clamping — is what preserves full coverage from a mid-block start.
    static func orderedTraversal(
        paths: [Double],
        startPath: Double,
        higherFirst: Bool
    ) -> [Double] {
        guard !paths.isEmpty else { return [] }
        let sorted = higherFirst ? paths.sorted() : paths.sorted(by: >)
        guard let idx = nearestIndex(in: sorted, to: startPath) else { return sorted }
        return Array(sorted[idx...]) + Array(sorted[..<idx])
    }

    /// Index of the path closest to `value`. The start path is clamped/snapped
    /// before it reaches here; this is the final guarantee that traversal
    /// begins on a real member of the universe.
    private static func nearestIndex(in ordered: [Double], to value: Double) -> Int? {
        guard let first = ordered.first else { return nil }
        var best = 0
        var bestDelta = abs(first - value)
        for (i, p) in ordered.enumerated() {
            let delta = abs(p - value)
            if delta < bestDelta {
                best = i
                bestDelta = delta
            }
        }
        return best
    }

    private static func indexed(_ ordered: [Double], by indices: [Int]) -> [Double] {
        indices.compactMap { $0 >= 0 && $0 < ordered.count ? ordered[$0] : nil }
    }

    /// 3/5 pattern in index space: startup 0, +3, +1, +5, +2 then alternating
    /// +5 / -3, finally sweeping up whatever the walk missed.
    private static func fiveThreeIndices(count: Int) -> [Int] {
        guard count > 0 else { return [] }
        var indices: [Int] = []
        var visited = Set<Int>()

        for offset in [0, 3, 1, 5, 2] where offset < count && !visited.contains(offset) {
            indices.append(offset)
            visited.insert(offset)
        }

        if var current = indices.last {
            var usePlusFive = true
            while true {
                let next = usePlusFive ? current + 5 : current - 3
                if next < 0 || next >= count || visited.contains(next) { break }
                indices.append(next)
                visited.insert(next)
                current = next
                usePlusFive.toggle()
            }
        }

        for i in 0..<count where !visited.contains(i) { indices.append(i) }
        return indices
    }

    /// 2 Row Up & Back in index space: out on every 4th path from offset 1,
    /// back on every 4th from offset 3, then the remainder in traversal order.
    private static func twoRowUpBackIndices(count: Int) -> [Int] {
        guard count > 0 else { return [] }
        var indices: [Int] = []
        var visited = Set<Int>()

        var i = 1
        while i < count {
            indices.append(i)
            visited.insert(i)
            i += 4
        }

        var back: [Int] = []
        i = 3
        while i < count {
            back.append(i)
            i += 4
        }
        for j in back.reversed() where !visited.contains(j) {
            indices.append(j)
            visited.insert(j)
        }

        for k in 0..<count where !visited.contains(k) { indices.append(k) }
        return indices
    }

    /// Every Second Row, parity-preserving sequence across the combined paths.
    static func everySecondRowSequence(
        paths: [Double],
        startPath: Double,
        higherFirst: Bool
    ) -> [Double] {
        guard !paths.isEmpty else { return [] }
        let sorted = paths.sorted()
        let sameParity = sorted.filter { p in
            let diff = Int(round(p - startPath))
            return diff % 2 == 0
        }
        guard !sameParity.isEmpty else { return [] }
        if higherFirst {
            let firstRun = sameParity.filter { $0 >= startPath }.sorted()
            let wrap = sameParity.filter { $0 < startPath }.sorted()
            return firstRun + wrap
        } else {
            let firstRun = sameParity.filter { $0 <= startPath }.sorted(by: >)
            let wrap = sameParity.filter { $0 > startPath }.sorted(by: >)
            return firstRun + wrap
        }
    }

    /// Short preview text for the Proposed Row Sequence card. Caps to the
    /// first `maxItems` and adds an ellipsis when truncated.
    static func sequencePreviewText(_ sequence: [Double], maxItems: Int = 10) -> String {
        let preview = sequence.prefix(maxItems).map { formatPath($0) }
        let joined = preview.joined(separator: " → ")
        return sequence.count > maxItems ? joined + " → …" : joined
    }

    /// Human-readable explanation of how a pattern walks the paths. `nil`
    /// for Free Drive (no planned sequence).
    static func patternPreviewNote(_ pattern: TrackingPattern) -> String? {
        switch pattern {
        case .sequential:
            return "Sequential: walks every path one-by-one in the chosen direction."
        case .everySecondRow:
            return "Every Second Row: advances by +2 in the chosen direction, then wraps to cover the remaining same-parity paths."
        case .fiveThree:
            return "5/3 pattern: skips ahead 5, back 3, repeating from the chosen start."
        case .upAndBack:
            return "Up and Back: traverses then reverses, covering each path once."
        case .twoRowUpBack:
            return "Two Row Up & Back: pairs of rows, advancing then returning."
        case .custom:
            return "Custom pattern: generated from the chosen start and direction."
        case .freeDrive:
            return nil
        }
    }
}
